src/org/skypro/skyshop/basket/ProductBasket.java

```diff
@@ -1,56 +1,49 @@
 package org.skypro.skyshop.basket;
 
-import org.skypro.skyshop.product.DiscountedProduct;
 import org.skypro.skyshop.product.Product;
-import org.skypro.skyshop.product.SimpleProduct;
+
+import java.util.*;
 
 public class ProductBasket {
-    private Product[] products = new Product[5];
+    private HashMap<String, List<Product>> products;
 
-    public ProductBasket(Product[] products) {
-        for (int i = 0; i < this.products.length; i++) {
-            if (i < products.length) {
-                this.products[i] = products[i];
-            }
+    public ProductBasket(List<Product> products) {
+        this.products = new HashMap<>();
+
+        for (Product p : products) {
+            List<Product> valList = new ArrayList<Product>();
+            valList.add(p);
+            this.products.put(p.getName(), valList);
         }
     }
 
     public void addProduct(Product product) {
-        boolean added = false;
-
-        for (int i = 0; i < products.length; i++) {
-            if (products[i] == null) {
-                products[i] = product;
-                added = true;
-                break;
-            }
-        }
+        ArrayList<Product> currVal = (ArrayList<Product>) products.putIfAbsent(product.getName(), List.of(product));
```

Comment:
Аналогично, абстракция - один из принципов ООП. Используем максимально возможный уровень абстракции, тогда и кастовать ничего не придется

```suggestion
        List<Product> currVal = products.putIfAbsent(product.getName(), List.of(product));

```